src/v/cluster/panda_link/frontend.cc

```diff
@@ -0,0 +1,297 @@
+        case raft::errc::timeout:
+            return cluster::errc::timeout;
+        case raft::errc::not_leader:
+            return cluster::errc::not_leader_controller;
+        default:
+            return cluster::errc::replication_error;
+        }
+    }
+    if (ec.category() == rpc::error_category()) {
+        switch (rpc::errc(ec.value())) {
+        case rpc::errc::client_request_timeout:
+            return cluster::errc::timeout;
+        default:
+            return cluster::errc::replication_error;
+        }
+    }
+    if (ec.category() == cluster::error_category()) {
+        return cluster::errc(ec.value());
+    }
+    return cluster::errc::replication_error;
+}
+} // namespace
+
+frontend::frontend(
+  model::node_id self,
+  cluster::partition_leaders_table* leaders,
+  table* table,
+  cluster::controller_stm* controller,
+  rpc::connection_cache* connections,
+  features::feature_table* features,
+  ss::abort_source* as)
+  : _self(self)
+  , _leaders(leaders)
+  , _connections(connections)
+  , _table(table)
+  , _as(as)
+  , _controller(controller)
+  , _features(features) {}
+
+ss::future<mutation_result> frontend::upsert_panda_link(
+  metadata meta, model::timeout_clock::time_point timeout) {
+    if (!panda_link_active(true)) {
+        co_return mutation_result{.ec = cluster::errc::feature_disabled};
+    }
+    panda_link_cmd c{cluster::panda_link_upsert_cmd{0, std::move(meta)}};
+    co_return co_await do_mutation(std::move(c), timeout);
+}
+
+ss::future<mutation_result> frontend::remove_panda_link(
+  name_t name, model::timeout_clock::time_point timeout) {
+    if (!panda_link_active(false)) {
+        co_return mutation_result{.ec = cluster::errc::feature_disabled};
+    }
+    panda_link_cmd c{cluster::panda_link_remove_cmd(std::move(name), 0)};
+    co_return co_await do_mutation(std::move(c), timeout);
+}
+
+bool frontend::panda_link_active(bool check_license) const {
+    return _features->is_active(features::feature::panda_linking_dr)
+           && !(check_license && _features->should_sanction());
+}
+
+ss::future<mutation_result> frontend::do_mutation(
+  panda_link_cmd cmd, model::timeout_clock::time_point timeout) {
+    auto cluster_leader = _leaders->get_leader(model::controller_ntp);
+    if (!cluster_leader) {
+        co_return mutation_result{.ec = cluster::errc::not_leader_controller};
+    }
+    if (*cluster_leader != _self) {
+        co_return co_await dispatch_mutation_to_remote(
+          *cluster_leader,
+          std::move(cmd),
+          timeout - model::timeout_clock::now());
+    }
+    if (ss::this_shard_id() != cluster::controller_stm_shard) {
+        co_return co_await container().invoke_on(
+          cluster::controller_stm_shard,
+          [cmd = std::move(cmd), timeout](auto& service) mutable {
+              return service.do_mutation(std::move(cmd), timeout);
+          });
+    }
+
+    co_return co_await do_local_mutation(std::move(cmd), timeout);
```

Comment:
looks like this can be simplified to: 
```
co_return co_await container().invoke_on(
    cluster::controller_stm_shard,
    [cmd = std::move(cmd), timeout](auto& service) mutable {
        return service.do_local_mutation(std::move(cmd), timeout);
    });
```